DEVELOPERS_GUIDE.md

```diff
@@ -0,0 +1,203 @@
+>*/
+
+### Formatting
+
+The code is formatted using [clang-format](https://clang.llvm.org/docs/ClangFormat.html) using the Google formatting conventions. The appropriate ``.clang-format`` is in the main directory. 
+
+
+### General Recommendations
+
+- Any violation to the guide is allowed if it enhances readability.
+
+### Naming Conventions
+
+- File names must be of the form ``FileName``.
+- Directory names must be of the form ``DirectoryName``.
+- Names representing types must be in mixed case starting with upper case, i.e. ``TypeName``.
+- Variable names must be in mixed case starting with lower case, i.e. ``variableName``.
+- Named constants (including enumeration values) must be all uppercase using underscore to separate words.
+- Names representing methods or functions must be verbs and written in mixed case starting with lower case, i.e. ``functionName``.
+- Names representing Classes must be of the form ``ClassName``.
+- Names representing template types should be a single uppercase letter.
+- Private class variables should have ``_`` suffix.
+- Never use identifiers which begin with one or two underscores (`_' or `__'). In most cases those variables are used in imported libraries.
+- Plural form should be used on names representing a collection of objects.
+- A variable with a large scope should have a long name.
+- Choose variable names that suggest the usage.
+- The prefix ``numberOf`` should be used for variables representing a number of objects.
+- The suffix ``Id`` should be used for variables representing an entity number.
+- The prefix ``is`` or ``do`` or ``use`` or conjugations thereof should be used for boolean variables and methods.
+- Negated boolean variable names must be avoided.
+
+### Files
+
+- C++ header files should have the extension ``.h``.
+- Source files should have the extension ``.cpp``.
+- Application files should have the extension ``.cpp``.
+- Implementation files of templated classes should have the extension ``_Impl.h`` and be included at the end of the header file.
+- Large source files can be split in multiple files by using the same name, then an ``_`` and a description of its usage.
+- A class should be declared in a header file and defined in a source file where the name of the files match the name of the class.
+- All definitions should reside in source files.
```

Comment:
This is violated so many times in the current state of the repository. Probably to allow the compiler to inline functions. I do not know a good alternative guideline at the moment.

Reply:
I think the best way is to not put it in the class declaration, but underneath in the same file. I agree that putting it in a separate header is not beneficial.